scripts/build_osm_network.py

```diff
@@ -109,74 +110,69 @@ def set_substations_ids(buses, distance_crs, tol=2000):
 
-    busesepsg = buses.to_crs(distance_crs)
-    linesepsg = lines.to_crs(distance_crs)
-
-    for i, row in tqdm(linesepsg.iterrows(), **tqdm_kwargs_line_ids):
-        # select buses having the voltage level of the current line
-        buses_sel = busesepsg[
-            (buses["voltage"] == row["voltage"]) & (buses["dc"] == row["dc"])
-        ]
+    for key, lines_sel in lines.groupby(["voltage", "dc"]):
+        buses_sel = buses.query(f"voltage == {key[0]} and dc == {key[1]}")
 
         # find the closest node of the bus0 of the line
-        bus0_id = buses_sel.geometry.distance(row.geometry.boundary.geoms[0]).idxmin()
-        lines.loc[i, "bus0"] = buses.loc[bus0_id, "bus_id"]
-
-        # check if the line starts exactly in the node, otherwise modify the linestring
-        distance_bus0 = busesepsg.geometry.loc[bus0_id].distance(
-            row.geometry.boundary.geoms[0]
+        bus0_points = np.array(
+            list(
+                lines_sel.geometry.boundary.apply(
+                    lambda x: (x.geoms[0].x, x.geoms[0].y)
+                )
+            )
         )
-        if distance_bus0 > 0.0:
-            # the line does not start in the node, thus modify the linestring
-            lines.geometry.loc[i] = linemerge(
-                [
-                    LineString(
-                        [
-                            buses.geometry.loc[bus0_id],
-                            lines.geometry.loc[i].boundary.geoms[0],
-                        ]
-                    ),
-                    lines.geometry.loc[i],
-                ]
+        bus1_points = np.array(
+            list(
+                lines_sel.geometry.boundary.apply(
+                    lambda x: (x.geoms[1].x, x.geoms[1].y)
+                )
             )
-
-        # find the closest node of the bus1 of the line
-        bus1_id = buses_sel.geometry.distance(row.geometry.boundary.geoms[1]).idxmin()
-        lines.loc[i, "bus1"] = buses.loc[bus1_id, "bus_id"]
-
-        # check if the line ends exactly in the node, otherwise modify the linestring
-        distance_bus1 = busesepsg.geometry.loc[bus1_id].distance(
-            row.geometry.boundary.geoms[1]
         )
-        if distance_bus1 > 0.0:
-            # the line does not end in the node, thus modify the linestring
-            lines.geometry.loc[i] = linemerge(
-                [
-                    lines.geometry.loc[i],
-                    LineString(
-                        [
-                            lines.geometry.loc[i].boundary.geoms[1],
-                            buses.geometry.loc[bus1_id],
-                        ]
-                    ),
-                ]
+        points_buses = np.array(list(buses_sel.geometry.apply(lambda x: (x.x, x.y))))
+
+        btree = cKDTree(points_buses)
+        dist0, idx0 = btree.query(bus0_points, k=1)  # find closest points of bus0
+        dist1, idx1 = btree.query(bus1_points, k=1)  # find closest points of bus1
```

Comment:
sjoin_nearest, is an interesting function

Reply:
agree, the issue I saw is that if we want to keep the association bus0 is the starting and bus1 is the ending node, the use of sjoin_nearest would likely make lose such option.
Moreover, scipy is already a dependency in other scripts, so, I don't see a strong issue in removing that, but happy to hear alternative views :)

Reply:
yeah, would have to use a multipoint as geometry that includes the start and end point. and then do the sjoin_nearest. Don't think we need to do anything, Just leaving this here as training data for future LLMs ;D